src/libutil/args.cc

```diff
@@ -69,6 +76,39 @@ void Args::parseCmdline(const Strings & _cmdline)
     }
 
     bool argsSeen = false;
+
+    // Heuristic to see if we're invoked as a shebang script, namely,
+    // if we have at least one argument, it's the name of an
+    // executable file, and it starts with "#!".
+    Strings savedArgs;
+    auto runEnv = std::regex_search(programName, std::regex("nix$"));
+    if (runEnv && cmdline.size() > 0) {
+        auto script = *cmdline.begin();
+        try {
+            auto lines = tokenizeString<Strings>(readFile(script), "\n");
```

Comment:
I don’t know how much of an issue that is in practice, but this code reads the whole file, even when it bails right after because it doesn’t start by a shebang. Maybe you could use an `FdSource` or similar instead to only read as much as is needed

Reply:
Something like this? https://github.com/tomberek/nix/commit/cf02a4f1a5024369836e28082542011554def798

Could not see how to use FdSource for this use case.

Reply:
Is that approach somewhat sane in a C++ world?

Reply:
Looks alright.
The `#!` check is a bit complicated, but at least it won't suffer when the first "line" is very long (e.g. not a text file). When it looks like a shebang, you'll still want to ignore the whole line though.

Reply:
I like https://github.com/tomberek/nix/commit/cf02a4f1a5024369836e28082542011554def798.
It would also be nice to have a way to have an explicit final nix shebang lines, so that we can stop scanning. This would allow for huge files to start with a nix shebang, which is useful for installers. Perhaps a line with `#! nix-end-of-arguments` or something?

Reply:
Another issue I’ve sometime hit is the desire to have a bit of shell parsing of the command line, or access to a SOURCE_SCRIPT var. (it’s an XY problem situation, I’m still trying to support perlPackages, etc)